SimG4CMS/Calo/src/HFNoseSD.cc

```diff
@@ -0,0 +1,236 @@
+///////////////////////////////////////////////////////////////////////////////
+// File: HFNoseSD.cc
+// Description: Sensitive Detector class for High Granularity Calorimeter
+///////////////////////////////////////////////////////////////////////////////
+
+#include "DataFormats/Math/interface/FastMath.h"
+#include "DataFormats/ForwardDetId/interface/HGCSiliconDetId.h"
+#include "SimG4CMS/Calo/interface/HFNoseSD.h"
+#include "SimG4Core/Notification/interface/TrackInformation.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/HGCalCommonData/interface/HGCalDDDConstants.h"
+#include "Geometry/HGCalCommonData/interface/HGCalGeometryMode.h"
+#include "G4LogicalVolumeStore.hh"
+#include "G4LogicalVolume.hh"
+#include "G4Step.hh"
+#include "G4Track.hh"
+#include "G4ParticleTable.hh"
+#include "G4VProcess.hh"
+#include "G4Trap.hh"
+
+#include <iostream>
+#include <fstream>
+#include <iomanip>
+
+//#define EDM_ML_DEBUG
+
+HFNoseSD::HFNoseSD(const std::string& name, const DDCompactView & cpv,
+		   const SensitiveDetectorCatalog & clg, 
+		   edm::ParameterSet const & p, const SimTrackManager* manager):
+  CaloSD(name, cpv, clg, p, manager,
+         (float)(p.getParameter<edm::ParameterSet>("HFNoseSD").getParameter<double>("TimeSliceUnit")),
+         p.getParameter<edm::ParameterSet>("HFNoseSD").getParameter<bool>("IgnoreTrackID")), 
+  hgcons_(nullptr), slopeMin_(0), levelT1_(99), levelT2_(99), 
+  tan30deg_(std::tan(30.0*CLHEP::deg)) {
+
+  numberingScheme_.reset(nullptr); mouseBite_.reset(nullptr);
+
+  edm::ParameterSet m_HGC = p.getParameter<edm::ParameterSet>("HFNoseSD");
+  eminHit_         = m_HGC.getParameter<double>("EminHit")*CLHEP::MeV;
+  fiducialCut_     = m_HGC.getParameter<bool>("FiducialCut");
+  distanceFromEdge_= m_HGC.getParameter<double>("DistanceFromEdge");
+  storeAllG4Hits_  = m_HGC.getParameter<bool>("StoreAllG4Hits");
+  rejectMB_        = m_HGC.getParameter<bool>("RejectMouseBite");
+  waferRot_        = m_HGC.getParameter<bool>("RotatedWafer");
+  angles_          = m_HGC.getUntrackedParameter<std::vector<double>>("WaferAngles");
+
+  if(storeAllG4Hits_) {
+    setUseMap(false);
+    setNumberCheckedHits(0);
+  }
+
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCSim")<< "**************************************************"
+			    << "\n"
+			    << "*                                                *"
+			    << "\n"
+			    << "* Constructing a HFNoseSD  with name " << name << "\n"
+			    << "*                                                *"
+			    << "\n"
+			    << "**************************************************";
+#endif
+  edm::LogVerbatim("HGCSim") << "HFNoseSD:: Threshold for storing hits: " 
```

Comment:
should these all be inside the `EDM_ML_DEBUG` ifdef?

Reply:
I prefer to keep it out of EDM_ML_DEBUG